File: wrappers/java/src/main/java/com/turbotoken/RankCache.java
package com.turbotoken;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Downloads and caches rank files for BPE encodings.
 * Cache directory: TURBOTOKEN_CACHE_DIR env or ~/.cache/turbotoken/
 */
final class RankCache {

    private RankCache() {}

    static Path getCacheDir() {
        String envDir = System.getenv("TURBOTOKEN_CACHE_DIR");
        if (envDir != null && !envDir.isEmpty()) {
            return Paths.get(envDir);
        }
        return Paths.get(System.getProperty("user.home"), ".cache", "turbotoken");
    }

    /**
     * Ensures the rank file for the given encoding name is cached locally.
     * Downloads it from the spec's URL if missing.
     * @return the path to the cached rank file
     */
    static Path ensureRankFile(String encodingName) {
        Registry.EncodingSpec spec = Registry.getEncodingSpec(encodingName);
        String fileName = encodingName + ".tiktoken";
        Path cacheDir = getCacheDir();
        Path cached = cacheDir.resolve(fileName);

        if (Files.exists(cached)) {
            return cached;
        }

        try {
            Files.createDirectories(cacheDir);

            HttpClient client = HttpClient.newHttpClient();
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(spec.getRankFileUrl()))
                .GET()
                .build();
            HttpResponse<byte[]> response = client.send(request, HttpResponse.BodyHandlers.ofByteArray());

            if (response.statusCode() != 200) {
                throw new TurboTokenException(
                    "Failed to download rank file for " + encodingName
                    + ": HTTP " + response.statusCode()
                );
            }

            Files.write(cached, response.body());
            return cached;

        } catch (IOException | InterruptedException e) {
            throw new TurboTokenException("Failed to download rank file for " + encodingName, e);
        }
    }

    /**
     * Reads the rank file bytes for the given encoding name,
     * downloading it first if not cached.
     */
    static byte[] readRankFile(String encodingName) {
        Path path = ensureRankFile(encodingName);
        try {
            return Files.readAllBytes(path);
        } catch (IOException e) {
            throw new TurboTokenException("Failed to read rank file: " + path, e);
        }
    }
}

File: flutter/lib/src/rank_cache.dart
/// Rank file downloading and caching.

import 'dart:io';
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';

import 'error.dart';
import 'registry.dart';

Future<Directory> cacheDir() async {
  final envDir = Platform.environment['TURBOTOKEN_CACHE_DIR'];
  if (envDir != null && envDir.isNotEmpty) {
    final dir = Directory(envDir);
    if (!await dir.exists()) {
      await dir.create(recursive: true);
    }
    return dir;
  }

  final appCache = await getApplicationCacheDirectory();
  final dir = Directory(p.join(appCache.path, 'turbotoken'));
  if (!await dir.exists()) {
    await dir.create(recursive: true);
  }
  return dir;
}

Future<File> ensureRankFile(String name) async {
  final spec = getEncodingSpec(name);
  final dir = await cacheDir();
  final file = File(p.join(dir.path, '$name.tiktoken'));

  if (await file.exists()) {
    return file;
  }

  final response = await http.get(Uri.parse(spec.rankFileUrl));
  if (response.statusCode != 200) {
    throw DownloadException(spec.rankFileUrl, statusCode: response.statusCode);
  }

  await file.writeAsBytes(response.bodyBytes);
  return file;
}

Future<Uint8List> readRankFile(String name) async {
  final file = await ensureRankFile(name);
  return file.readAsBytes();
}
